### ISSUE TITLE
Conversion of inputs in distancematrix

### ISSUE THREAD

**Issue by heliosdrm** [JuliaDynamics/RecurrenceAnalysis.jl#18]
**TL;DR:** I think we should revert that change introduced in a1d04043

The code of `distancematrix` was changed in a1d04043 and 86d7356 from:

    if sx[2] > _maxdimension
        return _distancematrix(Matrix(x), Matrix(y), metric)
    else
        return _distancematrix(Dataset(x), Dataset(y), metric)
    end

to:

    if sx[2] < MAXDIM # convert to Dataset here, it is significantly faster
        return _distancematrix(Dataset(x), Dataset(y), metric)
    else
        return _distancematrix(x, y, metric)
    end

Let aside the name of the constant (previously `_maxdimension`, changed to `MAXDIM`) and the sign of the comparison, the main difference is that when the dimension of the embedding space is greater equal than `MAXDIM`, `x` and `y` were converted into matrices to gain performance (cf. https://gist.github.com/Datseris/3ca61b32dea2033aeae5c4dfe3e0b78c), and now they are left as they were.

But in many (perhaps most) cases, if `x` or `y`  been made by `embed` or `reconstruct`, they won't be matrices, but `Dataset`s, so even although their dimension is greater than `MAXDIM`, the two branches of the condition will run the same method.

Or worse (although that's a far-fetched rare case): if one of them is a `Matrix` and the other a `Dataset`, and their dimension is greater than `MAXDIM`, no appropriate method of `_distancematrix` will be found and the function will fail.

**Comment by Datseris** [comment]
There is a reason I did this change: I benchmarked these things on the stable branch and I realized that whatever I had in this old Jupyter notebook, was not true anymore. I could not find a case where `Dataset` was slower than matrix. I only found `Dataset` to be _much faster_ than matrix, and only for high dimensions they were equal in speed!

I'll post these benchmark codes tomorrow morning ASAP!

**Comment by Datseris** [comment]
Alright, I am running these on my home computer, on current master branch:
```julia
using RecurrenceAnalysis, BenchmarkTools, Distances

dm = RecurrenceAnalysis._distancematrix

const metric = Chebyshev()

for N in [100, 1000]
    println("N=$N")
    x = rand(N)
    for γ in [3, 9, 25]
        println("γ=$γ")
        d = RecurrenceAnalysis.reconstruct(x, γ, 1)
        d2 = Matrix(d)
        @btime dm($d, $d, $metric)
        @btime dm($d2, $d2, $metric)
    end
end
```
with result:
```
N=100
γ=3
  65.277 μs (2 allocations: 73.64 KiB)
  90.320 μs (4 allocations: 73.67 KiB)
γ=9
  191.726 μs (2 allocations: 64.83 KiB)
  181.872 μs (4 allocations: 64.86 KiB)
γ=25
  378.936 μs (2 allocations: 44.08 KiB)
  336.239 μs (4 allocations: 44.11 KiB)
N=1000
γ=3
  9.345 ms (2 allocations: 7.58 MiB)
  11.806 ms (4 allocations: 7.58 MiB)
γ=9
  26.399 ms (2 allocations: 7.49 MiB)
  25.172 ms (4 allocations: 7.49 MiB)
γ=25
  65.176 ms (2 allocations: 7.25 MiB)
  65.632 ms (4 allocations: 7.25 MiB)
```

I can tell you for sure that for high dimensionality in both my laptops both methods (for significantly large N) have same speed. However it is always the case that for small dimensions the dataset method is significantly faster.

**Comment by heliosdrm** [comment]
I was trying at the same time. It's different if `metric=Euclidean()`: I don't know why, but the time increases with greater dimensions and `Dataset`, but it's about the same with `Matrix`, regardless of the dimension!

So if the metric is `Euclidean()`, it does make sense to convert.

**Comment by Datseris** [comment]
Ohhhhhhhhhhhhh daaaaamn you are right. Of course! The answer is here: https://discourse.julialang.org/t/pairwise-distance-using-matrix-or-vector-svector-the-first-is-faster/17312/2?u=datseris 

Distances uses some heavy BLAS optimizations for `Euclidean`. Okay, we have to take care of this then and indeed convert for `Euclidean` (or even any of the other distances we use and we have a performance gain). 

Sorry I overlooked this!

**Comment by heliosdrm** [comment]
Ok, that also explains why computing the Euclidean distance is significantly faster than the Chebyshev norm - except for very few dimensions. I was a bit surprised to see that.
